node/src/components/consensus/era_supervisor.rs

```diff
@@ -582,6 +581,87 @@ where
         }
     }
 
+    pub(super) fn handle_initialize_eras(
+        &mut self,
+        switch_blocks: HashMap<EraId, BlockHeader>,
+        current_era: EraId,
+        validators: BTreeMap<PublicKey, U512>,
+        genesis_state_root_hash: Digest,
+        timestamp: Timestamp,
+        genesis_start_time: Timestamp,
+    ) -> Effects<Event<I>> {
+        let last_upgrade_point = self.era_supervisor.protocol_config.last_upgrade_point;
+        let mut effects: Effects<Event<I>> = Default::default();
+
+        for era_id in
+            current_era.iter_bonded(last_upgrade_point, self.era_supervisor.bonded_eras * 2)
+        {
+            let maybe_switch_block = if era_id > last_upgrade_point {
+                Some(EraId(era_id.0 - 1))
+            } else {
+                None
+            }
+            .and_then(|switch_id| switch_blocks.get(&switch_id));
+
+            let newly_slashed = maybe_switch_block
+                .and_then(|bhdr| bhdr.era_end())
+                .map(|era_end| era_end.equivocators.clone())
+                .unwrap_or_default();
+
+            let slashed = era_id
+                .iter_other(last_upgrade_point, self.era_supervisor.bonded_eras)
+                .filter_map(|old_id| switch_blocks.get(&old_id).and_then(|bhdr| bhdr.era_end()))
+                .flat_map(|era_end| era_end.equivocators.clone())
+                .collect();
+
+            let booking_era_id = EraId(
+                era_id
+                    .0
+                    .saturating_sub(self.era_supervisor.protocol_config.auction_delay + 1),
+            );
+
+            let seed = if let Some(booking_block) = switch_blocks.get(&booking_era_id) {
+                EraSupervisor::<I>::era_seed(booking_block.hash(), booking_block.accumulated_seed())
+            } else {
+                0
+            };
+
+            let (validators, start_height, state_root_hash) =
+                if let Some(switch_block) = maybe_switch_block {
+                    (
+                        switch_block
+                            .next_era_validator_weights()
+                            .cloned()
+                            .expect("switch block should have era validator weights"),
+                        switch_block.height() + 1,
+                        *switch_block.state_root_hash(),
+                    )
+                } else {
+                    (validators.clone(), 0, genesis_state_root_hash)
```

Comment:
This shouldn't be `0, genesis_state_root_hash` for upgrade points, though, should it?

Reply:
But for upgrade points, we should have the previous switch blocks, which means we shouldn't land in this branch. Unless it's an emergency restart... Hmm, I'm not sure here.

Reply:
But you set `maybe_switch_block` to `None` unless `era_id > last_upgrade_point` — should that be a `>=` above?

Reply:
Ah, good point. So I guess there is a design decision to be made here about what _should_ we do for the first era after an upgrade.

I think it will be safer to _not_ use the previous era's switch block for two reasons:
- The upgrade might in fact be an emergency restart and there is _no_ previous switch block,
- The upgrade might contain changes to the validator set and other things, making the values from the switch block invalid.

I'm not sure if the latter can happen outside of an emergency restart, but I guess it doesn't hurt to cater to such a use case. If we go down this route, we'll have to know the state hash and the block height associated with every upgrade, though.

Reply:
I think both of your points are valid and can coexist. We shouldn't be guessing but that information should be part of the startup configuration.

Reply:
To my mind there is only one right answer:
* If it is an emergency restart — i.e. if the upgrade point contains an explicit block hash and validator set — we must use the validator set specifies in the upgrade point.
* If it's a regular upgrade we must use the pre-upgrade switch block. (What else would we use? We can't start at height 0 and with genesis after every upgrade?)

Reply:
I agree. Although, an emergency restart not necessarily has to change the validator set. Then we should be using validators from the switch block that "described" the era in which the explicit block hash lives in.

Reply:
But then one switch block will be missing.

Reply:
I don't understand. I basically modified your first bullet-point so that it does not require validator set.

Reply:
I mean after an emergency restart there will be one era without a switch block. So one future era will be missing its booking block.

Reply:
So I think it fhis is an upgrade, it should be the same as in the first case, i.e. use the switch block. If it is an emergency restart, it should be:
```suggestion
                    (validators.clone(), last_height_before_crash, modified_root_hash)
```
where `modified_root_hash` is computed based on the root hash from the last block before the crash, by applying the specified changes.

If this PR doesn't support these cases yet, I'd add `TODO`s here.

Reply:
I'd leave a TODO. The problem is that with the current state of the code we can't really get the necessary data (like the last block height or the modified root hash) - this stuff is still not implemented and only will be in the near future.